bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/UpdateEspIdfHandler.java

```diff
@@ -0,0 +1,109 @@
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.core.commands.AbstractHandler;
+import org.eclipse.core.commands.ExecutionEvent;
+import org.eclipse.core.commands.ExecutionException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.ui.console.MessageConsoleStream;
+
+import com.espressif.idf.core.ExecutableFinder;
+import com.espressif.idf.core.IDFEnvironmentVariables;
+import com.espressif.idf.core.logging.Logger;
+import com.espressif.idf.core.util.IDFUtil;
+import com.espressif.idf.ui.IDFConsole;
+
+public class UpdateEspIdfHandler extends AbstractHandler
+{
+
+	private MessageConsoleStream console;
+	private String gitExecutablePath;
+
+	@Override
+	public Object execute(ExecutionEvent event) throws ExecutionException
+	{
+		console = new IDFConsole().getConsoleStream(Messages.UpdateEspIdfCommand_Title, null);
+		IPath gitPath = ExecutableFinder.find("git", true); //$NON-NLS-1$
+		if (gitPath != null)
+		{
+			this.gitExecutablePath = gitPath.toOSString();
+
+		}
+		Thread updateEspIdfMasterThread = new Thread(new Runnable()
+		{
+			@Override
+			public void run()
+			{
+				ProcessBuilder[] builders = { new ProcessBuilder(getCheckoutMasterCommand()),
+						new ProcessBuilder(getGitPullCommand()), new ProcessBuilder(getSubmoduleUpdateCommand()) };
+				for (ProcessBuilder builder : builders)
+				{
+					builder.directory(new File(IDFUtil.getIDFPath()));
+					builder.redirectErrorStream(true);
+					Map<String, String> envMap = new IDFEnvironmentVariables().getEnvMap();
+					builder.environment().putAll(envMap);
+					startProcess(builder);
+				}
+
+			}
+
+			private void startProcess(ProcessBuilder builder)
+			{
+				try
+				{
+					Process process = builder.start();
+					InputStreamReader isr = new InputStreamReader(process.getInputStream(), StandardCharsets.UTF_8);
+					BufferedReader br = new BufferedReader(isr);
+					br.lines().forEach(line -> console.println((line)));
+				}
+				catch (IOException e)
+				{
+					Logger.log(e);
+				}
+			}
+		});
+		updateEspIdfMasterThread.start();
+		return null;
+	}
+
+	private List<String> getCheckoutMasterCommand()
+	{
+		List<String> command = new ArrayList<>();
+		command.add(gitExecutablePath);
+		command.add("checkout"); //$NON-NLS-1$
+		command.add("-f"); //$NON-NLS-1$
+		command.add("master"); //$NON-NLS-1$
+		return command;
+	}
+
+	private List<String> getGitPullCommand()
+	{
+		List<String> command = new ArrayList<>();
+		command.add(gitExecutablePath);
+		command.add("pull"); //$NON-NLS-1$
+		return command;
+	}
+
+	private List<String> getSubmoduleUpdateCommand()
+	{
+		List<String> command = new ArrayList<>();
+		command.add(gitExecutablePath);
+		command.add("submodule"); //$NON-NLS-1$
+		command.add("update"); //$NON-NLS-1$
+		command.add("--init"); //$NON-NLS-1$
+		command.add("--recursive"); //$NON-NLS-1$
+		return command;
+	}
```

Comment:
I think eclipse's own [org.eclipse.jgit.api.Git](https://download.eclipse.org/jgit/site/6.2.0.202206071550-r/apidocs/org/eclipse/jgit/api/package-summary.html) offers some better ways to do what we are trying to do manually here. You also can implement a progress monitor for that and it also provides a better way to work with the reporting on console. See if you can find anything in GitRepositoryBuilder.java class in our project in method public void repositoryClone(), although that particular implementation only takes care of cloning but seeing a few things in the documentation above you can see some methods that offer us other functionality like Pull. See this https://download.eclipse.org/jgit/site/6.2.0.202206071550-r/apidocs/org/eclipse/jgit/api/PullCommand.html